File: cos-runtime/rust/src/policy.rs
//! Permission helper — every claw-os bundled app or desktop binary
//! that touches a gated verb calls [`require`] before performing the
//! side effect. Mirrors the Python `cos_runtime.policy` module so
//! apps written in either language behave identically.
//!
//! ```no_run
//! use cos_runtime::policy;
//!
//! fn handle_rm(path: &str) -> Result<(), Box<dyn std::error::Error>> {
//!     policy::require("fs.delete", policy::Scope::path(path))?;
//!     std::fs::remove_file(path)?;
//!     Ok(())
//! }
//! ```
//!
//! [`require`] returns `Err(PolicyError::Denied(decision))` when the
//! kernel refuses; the inner [`Decision`] is the wire v1 perms reply.

use std::ffi::OsString;

use serde::{Deserialize, Serialize};

use crate::{cos_call_json, BridgeError};

/// Decision envelope returned by the hidden policy bridge. See
/// `wire/v1/perms.schema.json`.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Decision {
    pub decision: String,
    pub verb: String,
    #[serde(default)]
    pub scope: Option<serde_json::Value>,
    #[serde(default)]
    pub reason: Option<String>,
    #[serde(default)]
    pub hint: Option<String>,
    #[serde(default)]
    pub granted: Option<bool>,
}

impl Decision {
    pub fn is_allow(&self) -> bool {
        self.decision == "allow"
    }
}

/// Errors returned by [`require`] and [`check`].
#[derive(Debug, thiserror::Error)]
pub enum PolicyError {
    /// The kernel said no.
    #[error("permission denied: {} {:?}", .0.verb, .0.reason)]
    Denied(Decision),

    /// We couldn't even reach the kernel.
    #[error("policy unavailable: {0}")]
    Unavailable(String),

    /// Transport failure (cos binary, JSON decode, …).
    #[error(transparent)]
    Bridge(#[from] BridgeError),
}

/// Scope passed to [`require`] / [`check`]. Exactly one variant
/// corresponds to one `--path` / `--host` / `--name` / `--self` /
/// `--wild` flag on the hidden policy bridge.
#[derive(Debug, Clone)]
pub enum Scope {
    Path(String),
    Host(String),
    Name(String),
    SelfRef(String),
    Wild,
    Unscoped,
}

impl Scope {
    pub fn path(p: impl Into<String>) -> Self { Scope::Path(p.into()) }
    pub fn host(h: impl Into<String>) -> Self { Scope::Host(h.into()) }
    pub fn name(n: impl Into<String>) -> Self { Scope::Name(n.into()) }
    pub fn self_ref(s: impl Into<String>) -> Self { Scope::SelfRef(s.into()) }

    fn argv(&self) -> Vec<OsString> {
        match self {
            Scope::Path(v)    => vec!["--path".into(), v.into()],
            Scope::Host(v)    => vec!["--host".into(), v.into()],
            Scope::Name(v)    => vec!["--name".into(), v.into()],
            Scope::SelfRef(v) => vec!["--self".into(), v.into()],
            Scope::Wild       => vec!["--wild".into()],
            Scope::Unscoped   => vec![],
        }
    }
}

/// Ask the kernel whether the current process is allowed to perform
/// `verb` against `scope`. Returns `Ok(())` if allowed,
/// `Err(PolicyError::Denied(..))` if not.
pub fn require(verb: &str, scope: Scope) -> Result<(), PolicyError> {
    let decision = check(verb, scope)?;
    if decision.is_allow() {
        Ok(())
    } else {
        Err(PolicyError::Denied(decision))
    }
}

/// Same as [`require`] but returns the [`Decision`] envelope verbatim
/// — handy when the app wants to surface a "would-be-denied" hint
/// without aborting.
pub fn check(verb: &str, scope: Scope) -> Result<Decision, PolicyError> {
    let mut argv: Vec<OsString> = vec!["__policy".into(), "check".into(), verb.into()];
    argv.extend(scope.argv());

    let value = cos_call_json("policy", verb, argv).map_err(PolicyError::from)?;

    if value.get("decision").is_none() {
        return Err(PolicyError::Unavailable(format!(
            "policy check returned unrecognised envelope: {value}"
        )));
    }

    serde_json::from_value(value.clone()).map_err(|e| {
        PolicyError::Unavailable(format!("decision decode failed ({e}): {value}"))
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn scope_argv_shapes() {
        assert_eq!(Scope::Path("/etc".into()).argv(), &[OsString::from("--path"), OsString::from("/etc")]);
        assert_eq!(Scope::Wild.argv(), &[OsString::from("--wild")]);
        assert!(Scope::Unscoped.argv().is_empty());
    }

    #[test]
    fn decision_is_allow_helper() {
        let d = Decision { decision: "allow".into(), verb: "fs.read".into(), scope: None, reason: None, hint: None, granted: Some(true) };
        assert!(d.is_allow());
        let d = Decision { decision: "deny".into(), verb: "fs.read".into(), scope: None, reason: None, hint: None, granted: Some(false) };
        assert!(!d.is_allow());
    }
}

File: cos-runtime/python/src/cos_runtime/policy.py
"""Permission helper for Claw OS Python apps.

Every Python app should import this module and call ``policy.require()``
at the top of every operation that touches files, the network,
secrets, or anything else the kernel knows how to gate. The helper
shells out to the hidden policy bridge — the kernel's authoritative
enforcement entry point — so the answer here is exactly the same as
the answer the Rust side would give.

Typical usage::

    from cos_runtime import policy

    def handle_rm(args):
        policy.require("fs.delete", path=args["path"])
        os.remove(args["path"])

If the user has not granted the requested capability, ``require()``
raises :class:`PermissionDenied`, which the app's top-level driver
turns into a structured JSON error for the agent.

Why a subprocess and not an in-process check?
---------------------------------------------

The Python app already runs inside a process the kernel spawned with
``COS_SESSION`` set, so the subprocess call inherits the session and
PID-ancestry context the kernel needs to validate the request.
Centralising the decision in the hidden policy bridge keeps the Python
helper tiny, removes any risk of the rules drifting between Rust and
Python, and gives audit / logging a single chokepoint.
"""

from __future__ import annotations

import os
import shutil
import subprocess
from typing import Any, Mapping, Optional

from claw_os_sdk._transport import (
    WIRE_ARG,
    WireDenied,
    WireUnavailable,
    decode_response,
)


# Subprocess timeout for every shell-out to the hidden policy bridge. A hung
# child translates to PolicyUnavailable, which the caller can decide
# to treat as deny or warn-and-continue.
_DEFAULT_TIMEOUT_S = 60


# ---------------------------------------------------------------------------
# Exceptions
# ---------------------------------------------------------------------------


class PolicyError(Exception):
    """Base class for every error this module raises."""


class PermissionDenied(PolicyError):
    """The kernel refused the requested capability.

    ``denial`` holds the structured envelope returned by
    policy bridge — verb, requested scope, granted scopes,
    reason, hint — and is suitable for forwarding straight back to
    the caller as JSON.
    """

    def __init__(self, denial: Mapping[str, Any]):
        self.denial = dict(denial)
        super().__init__(self.denial.get("summary") or "permission denied")


class PolicyUnavailable(PolicyError):
    """The ``cos`` binary could not be invoked or returned garbage.

    This is distinct from :class:`PermissionDenied` so callers can
    decide whether a missing kernel implies "deny" (production) or
    "warn and continue" (dev tooling).
    """


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def require(
    verb: str,
    *,
    path: Optional[str] = None,
    host: Optional[str] = None,
    name: Optional[str] = None,
    self_ref: Optional[str] = None,
    wild: bool = False,
) -> None:
    """Check whether the current session may exercise ``verb``.

    Exactly one of ``path`` / ``host`` / ``name`` / ``self_ref`` /
    ``wild`` should be supplied. Unscoped verbs (``ui.notify``,
    ``time.delay`` …) take no keyword argument.

    Raises :class:`PermissionDenied` on deny, :class:`PolicyUnavailable`
    when the kernel cannot be reached. Returns ``None`` on allow.
    """
    decision = check(
        verb,
        path=path,
        host=host,
        name=name,
        self_ref=self_ref,
        wild=wild,
    )
    if decision.get("decision") != "allow":
        raise PermissionDenied(decision)


def check(
    verb: str,
    *,
    path: Optional[str] = None,
    host: Optional[str] = None,
    name: Optional[str] = None,
    self_ref: Optional[str] = None,
    wild: bool = False,
) -> dict:
    """Same as :func:`require` but returns the raw envelope instead of
    raising. Useful when an app wants to surface a "would-be-denied"
    notice without aborting.
    """
    cmd = [_cos_binary(), WIRE_ARG, "__policy", "check", verb]
    cmd.extend(_scope_flag(path=path, host=host, name=name, self_ref=self_ref, wild=wild))

    try:
        proc = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            check=False,
            timeout=_DEFAULT_TIMEOUT_S,
        )
    except subprocess.TimeoutExpired as exc:
        raise PolicyUnavailable(
            f"policy check timed out after {_DEFAULT_TIMEOUT_S}s"
        ) from exc

    try:
        decision = decode_response(
            (proc.stdout or "").strip(),
            proc.returncode,
            "policy check",
        )
    except (WireDenied, WireUnavailable) as error:
        raise PolicyUnavailable(str(error)) from error

    if "decision" not in decision:
        raise PolicyUnavailable("policy check response omitted `decision`")
    return decision


# ---------------------------------------------------------------------------
# Internals
# ---------------------------------------------------------------------------


def _scope_flag(
    *,
    path: Optional[str],
    host: Optional[str],
    name: Optional[str],
    self_ref: Optional[str],
    wild: bool,
) -> list:
    supplied = [
        ("--path", path),
        ("--host", host),
        ("--name", name),
        ("--self", self_ref),
    ]
    supplied = [(flag, val) for flag, val in supplied if val is not None]
    if wild:
        supplied.append(("--wild", None))
    if len(supplied) > 1:
        raise TypeError(
            "require()/check() accept at most one of path / host / name / self_ref / wild"
        )
    if not supplied:
        return []
    flag, val = supplied[0]
    return [flag] if val is None else [flag, val]


def _cos_binary() -> str:
    """Locate the ``cos`` binary. Honours ``CLAW_COS_BIN`` for tests."""
    override = os.environ.get("CLAW_COS_BIN")
    if override:
        return override
    found = shutil.which("cos")
    if found is None:
        raise PolicyUnavailable(
            "the `cos` binary is not on PATH; cannot enforce permissions"
        )
    return found
